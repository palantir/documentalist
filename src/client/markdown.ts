/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { IBlock } from "./compiler";
import { INavigable } from "./tags";

/**
 * The `MarkdownPlugin` exports a map of markdown `pages` keyed by reference,
 * and a multi-rooted `nav` tree of page nodes.
 */
export interface IMarkdownPluginData {
    /**
     * An ordered, nested, multi-rooted tree describing the navigation layout
     * of all the pages and their headings. The representation is constructued by
     * tracing `@page` and `@#+` tags.
     */
    nav: IPageNode[];

    /** A map of page reference to data. */
    pages: {
        [reference: string]: IPageData;
    };
}

/**
 * A single Documentalist page, parsed from a single source file.
 */
export interface IPageData extends IBlock {
    /** Unique identifier for addressing this page. */
    reference: string;

    /** Fully qualified route to this page: slash-separated references of all parent pages. */
    route: string;

    /** Human-friendly title of this page. */
    title: string;
}

/** An `@#+` tag belongs to a specific page. */
export interface IHeadingNode extends INavigable {
    /** Display title of page heading. */
    title: string;
}

/** A page has ordered children composed of `@#+` and `@page` tags. */
export interface IPageNode extends IHeadingNode {
    /** Ordered list of pages and headings that appear on this page. */
    children: Array<IPageNode | IHeadingNode>;

    /** Unique reference of this page, used for retrieval from store. */
    reference: string;
}

/** Type guard for `IPageNode`, useful for its `children` array. */
export function isPageNode(node: any): node is IPageNode {
    return node != null && (node as IPageNode).children != null;
}
