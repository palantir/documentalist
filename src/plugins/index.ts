/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

export * from "./kss";
export * from "./markdown";
export * from "./typescript";
export * from "./plugin";
