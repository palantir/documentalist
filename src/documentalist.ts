/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import * as fs from "fs";
import * as glob from "glob";
import * as path from "path";
import { IFile, IPlugin } from "./client";
import { Compiler, ICompilerOptions } from "./compiler";

/**
 * Plugins are stored with the regex used to match against file paths.
 */
export interface IPluginEntry<T> {
    pattern: RegExp;
    plugin: IPlugin<T>;
}

export class Documentalist<T> {
    /**
     * Construct a new `Documentalist` instance with the provided options.
     * This method lends itself particularly well to the chaining syntax:
     * `Documentalist.create(options).use(...)`.
     */
    public static create(options?: ICompilerOptions): Documentalist<{}> {
        return new Documentalist(options, []);
    }

    constructor(private options: ICompilerOptions = {}, private plugins: Array<IPluginEntry<T>> = []) {}

    /**
     * Adds the plugin to Documentalist. Returns a new instance of Documentalist
     * with a template type that includes the data from the plugin. This way the
     * `documentFiles` and `documentGlobs` methods will return an object that is
     * already typed to include each plugin's output.
     *
     * The plugin is applied to all files whose absolute path matches the
     * supplied pattern.
     *
     * @param pattern - A regexp pattern or a file extension string like "js"
     * @param plugin - The plugin implementation
     * @returns A new instance of `Documentalist` with an extended type
     */
    public use<P>(pattern: RegExp | string, plugin: IPlugin<P>): Documentalist<T & P> {
        if (typeof pattern === "string") {
            pattern = new RegExp(`${pattern}$`);
        }

        const newPlugins = [...this.plugins, { pattern, plugin }] as Array<IPluginEntry<T & P>>;
        return new Documentalist(this.options, newPlugins);
    }

    /**
     * Returns a new instance of Documentalist with no plugins.
     */
    public clearPlugins(): Documentalist<{}> {
        return new Documentalist<{}>(this.options, []);
    }

    /**
     * Finds all files matching the provided variadic glob expressions and then
     * runs `documentFiles` with them, emitting all the documentation data.
     */
    public async documentGlobs(...filesGlobs: string[]) {
        const files = this.expandGlobs(filesGlobs);
        return this.documentFiles(files);
    }

    /**
     * Iterates over all plugins, passing all matching files to each in turn.
     * The output of each plugin is merged to produce the resulting
     * documentation object.
     *
     * The return type `T` is a union of each plugin's data type.
     */
    public async documentFiles(files: IFile[]) {
        const compiler = new Compiler(this.options);
        // need an empty object of correct type that we can merge into
        // tslint:disable-next-line:no-object-literal-type-assertion
        const documentation = {} as T;
        for (const { pattern, plugin } of this.plugins) {
            const pluginFiles = files.filter(f => pattern.test(f.path));
            const pluginDocumentation = await plugin.compile(pluginFiles, compiler);
            this.mergeInto(documentation, pluginDocumentation);
        }
        return documentation;
    }

    /**
     * Expands an array of globs and flatten to a single array of files.
     */
    private expandGlobs(filesGlobs: string[]): IFile[] {
        return filesGlobs
            .map(filesGlob => glob.sync(filesGlob))
            .reduce((a, b) => a.concat(b))
            .map(fileName => {
                const absolutePath = path.resolve(fileName);
                return {
                    path: absolutePath,
                    read: () => fs.readFileSync(absolutePath, "utf8"),
                };
            });
    }

    /**
     * Shallow-merges keys form source into destination object (modifying it in the process).
     */
    private mergeInto(destination: T, source: T) {
        for (const key in source) {
            if (source.hasOwnProperty(key)) {
                if (destination.hasOwnProperty(key)) {
                    console.warn(`WARNING: Duplicate plugin key "${key}". Your plugins are overwriting each other.`);
                }
                destination[key] = source[key];
            }
        }
        return destination;
    }
}
