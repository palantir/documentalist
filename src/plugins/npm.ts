/**
 * Copyright 2018-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { spawn } from "child_process";
import { ICompiler, IFile, INpmPackage, INpmPluginData, IPlugin } from "../client";

export interface INpmPluginOptions {
    /** Whether to exclude packages marked `private`. */
    excludePrivate?: boolean;

    /** Array of patterns to exclude packages by `name`. */
    excludeNames?: Array<string | RegExp>;
}

/**
 * The `NpmPlugin` parses `package.json` files and emits data about each NPM
 * package. It uses `npm info` to look up data about published packages, and
 * falls back to `package.json` info if the package is private or unpublished.
 */
export class NpmPlugin implements IPlugin<INpmPluginData> {
    public constructor(private options: INpmPluginOptions = {}) {}

    public async compile(packageJsons: IFile[], dm: ICompiler): Promise<INpmPluginData> {
        const info = await Promise.all(packageJsons.map(pkg => this.parseNpmInfo(pkg, dm)));
        const { excludeNames, excludePrivate } = this.options;
        const npm = arrayToObject(
            info.filter(pkg => isNotExcluded(excludeNames, pkg.name) && excludePrivate !== pkg.private),
            pkg => pkg.name,
        );
        return { npm };
    }

    private parseNpmInfo = async (packageJson: IFile, dm: ICompiler): Promise<INpmPackage> => {
        const sourcePath = dm.relativePath(packageJson.path);
        const json = JSON.parse(packageJson.read());
        const data = JSON.parse(await this.getNpmInfo(json.name));
        // `npm info` returns an error if it doesn't know the package
        // so we can use package.json data instead
        if (data.error) {
            return {
                name: json.name,
                published: false,
                // tslint:disable-next-line:object-literal-sort-keys
                description: json.description,
                latestVersion: json.version,
                private: json.private === true,
                sourcePath,
                versions: [json.version],
            };
        }

        const distTags = data["dist-tags"] || {};
        return {
            name: data.name,
            published: true,
            // tslint:disable-next-line:object-literal-sort-keys
            description: data.description,
            latestVersion: distTags.latest,
            nextVersion: distTags.next,
            private: false,
            sourcePath,
            versions: data.versions,
        };
    };

    private getNpmInfo(packageName: string) {
        return new Promise<string>(resolve => {
            let stdout = "";
            const child = spawn("npm", ["info", "--json", packageName]);
            child.stdout.setEncoding("utf8");
            child.stdout.on("data", data => (stdout += data));
            child.on("close", () => resolve(stdout));
        });
    }
}

function arrayToObject<T>(array: T[], keyFn: ((item: T) => string)) {
    const obj: { [key: string]: T } = {};
    array.forEach(item => (obj[keyFn(item)] = item));
    return obj;
}

/** Returns true if value does not match all patterns. */
function isNotExcluded(patterns: Array<string | RegExp> = [], value?: string) {
    return value === undefined || patterns.every(p => value.match(p) == null);
}
