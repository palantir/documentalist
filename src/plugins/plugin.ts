/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { IBlock } from "../client";

/**
 * Abstract representation of a file, containing absolute path and synchronous `read` operation.
 */
export interface IFile {
    path: string;
    read(): string;
}

/**
 * Each plugin receives a `Compiler` instance to aid in the processing of source files.
 */
export interface ICompiler {
    /**
     * Converts an array of entries into a map of key to entry, using given
     * callback to extract key from each item.
     */
    objectify<T>(array: T[], getKey: (item: T) => string): { [key: string]: T };

    /**
     * Render a block of content by extracting metadata (YAML front matter) and
     * splitting text content into markdown-rendered HTML strings and `{ tag,
     * value }` objects.
     *
     * To prevent special strings like "@include" from being parsed, a reserved
     * tag words array may be provided, in which case the line will be left as
     * is.
     */
    renderBlock(blockContent: string, reservedTagWords?: string[]): IBlock;

    /**
     * Render a string of markdown to HTML, using the options from `Documentalist`.
     */
    renderMarkdown(markdown: string): string;
}

export interface IPlugin<T> {
    compile(files: IFile[], doc: ICompiler): T | Promise<T>;
}
