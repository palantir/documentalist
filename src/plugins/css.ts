/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import * as postcss from "postcss";
import { Comment, Root, Rule } from "postcss";
import * as postcssScss from "postcss-scss";

import { Documentalist } from "..";
import { StringOrTag } from "../client";
import { IFile, IPlugin } from "./plugin";

export interface IDeclaration {
    prop: string;
    value: string;
}

export interface IRule {
    comment?: StringOrTag[];
    commentRaw?: string;
    selector: string;
    declarations: IDeclaration[];
    metadata?: any;
}

export interface ICss {
    filePath: string;
    rules: IRule[];
}

export interface ICssPluginData {
    css: ICss[];
}

export class CssPlugin implements IPlugin<ICssPluginData> {

    public compile(documentalist: Documentalist<ICssPluginData>, cssFiles: IFile[]) {
        const css = [] as ICss[];
        cssFiles.forEach((file) => {
            const cssContent = file.read();
            const cssResult = {filePath: file.path, rules: []};
            postcss([this.processor(documentalist, cssResult)])
                .process(cssContent, { syntax: postcssScss })
                .css; // this statement makes the whole thing synchronous
            css.push(cssResult);
        });
        return { css };
    }

    private processor(documentalist: Documentalist<ICssPluginData>, cssResult: ICss) {
        return (css: Root) => {
            css.walkRules((rule: Rule) => {
                const ruleResult = {
                    declarations: [],
                    selector: rule.selector,
                } as IRule;

                const prevNode = rule.prev();
                if (prevNode != null && prevNode.type === "comment") {
                    const block = (prevNode as Comment).text
                        .replace(/^ *\*+ */, "")
                        .replace(/\n *\*+ */g, "\n")
                        .trim(); // trim asterisks, maintain newlines

                    const { content, metadata, renderedContent } = documentalist.renderBlock(block);
                    ruleResult.metadata = metadata;
                    ruleResult.commentRaw = content;
                    ruleResult.comment = renderedContent;
                }

                rule.walkDecls(({ prop, value }) => ruleResult.declarations.push({ prop, value }));

                cssResult.rules.push(ruleResult);
            });
        };
    }
}
