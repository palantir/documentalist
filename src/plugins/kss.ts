/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import * as kss from "kss";
import * as path from "path";
import { ICompiler, IFile, IPlugin } from "./plugin";

/** A single modifier for an example. */
export interface IKssModifier {
    documentation: string;
    name: string;
}

/**
 * A markup/modifiers example parsed from a KSS comment block.
 */
export interface IKssExample {
    /** Raw documentation string. */
    documentation: string;
    /** HTML markup for example, with `{{.modifier}}` templates. */
    markup: string;
    /** Array of modifiers supported by HTML markup. */
    modifiers: IKssModifier[];
    /** Unique reference for addressing this example. */
    reference: string;
}

export interface IKssPluginData {
    css: {
        [reference: string]: IKssExample;
    };
}

export class KssPlugin implements IPlugin<IKssPluginData> {
    public constructor(private options: kss.IOptions) {
    }

    public compile(cssFiles: IFile[], { objectify }: ICompiler) {
        const styleguide = this.parseFiles(cssFiles);
        const sections = styleguide.sections().map(convertSection);
        const css = objectify(sections, (s) => s.reference);
        return { css };
    }

    private parseFiles(files: IFile[]) {
        const input = files.map<kss.IFile>((file) => ({
            base: path.dirname(file.path),
            contents: file.read(),
            path: file.path,
        }));
        const options = { multiline: false, markdown: false, ...this.options };
        return kss.parse(input, options);
    }
}

function convertSection(section: kss.ISection): IKssExample {
    return {
        documentation: section.description(),
        markup: section.markup() || "",
        modifiers: section.modifiers().map(convertModifier),
        reference: section.reference(),
    };
}

function convertModifier(mod: kss.IModifier): IKssModifier {
    return {
        documentation: mod.description(),
        name: mod.name(),
    };
}
