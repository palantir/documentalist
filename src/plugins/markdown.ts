/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { IPageData, StringOrTag } from "../client";
import { makePage } from "../page";
import { ICompiler, IFile, IPlugin } from "./plugin";

export interface IMarkdownPluginData {
    docs: {
        [key: string]: IPageData;
    };
}

export class MarkdownPlugin implements IPlugin<IMarkdownPluginData> {
    /**
     * Reads the given set of markdown files and adds their data to the internal storage.
     * Returns a plain object mapping page references to their data.
     */
    public compile(markdownFiles: IFile[], { renderBlock }: ICompiler) {
        const pageStore: Map<string, IPageData> = new Map();
        markdownFiles
            .map((file) => {
                const absolutePath = file.path;
                const fileContents = file.read();
                const { content, metadata, renderedContent } = renderBlock(fileContents);
                const page = makePage({
                    absolutePath,
                    contentRaw: content,
                    contents: renderedContent,
                    metadata,
                });
                const ref = page.reference;
                if (pageStore.has(ref)) {
                    console.warn(`Found duplicate reference "${ref}"; overwriting previous data.`);
                    console.warn("Rename headings or use metadata `reference` key to disambiguate.");
                }
                pageStore.set(ref, page);
                return page;
            })
            .map((page) => {
                if (page.contents) {
                    const newContent = page.contents.reduce((array, content) => {
                        if (typeof content === "string" || content.tag !== "include") {
                            return array.concat(content);
                        }
                        const pageToInclude = pageStore.get(content.value as string);
                        if (pageToInclude === undefined) {
                            throw new Error(`Unknown @include reference '${content.value}' in '${page.reference}'`);
                        }
                        return array.concat(pageToInclude.contents!);
                    }, [] as StringOrTag[]);
                    page.contents = newContent;
                }
                return page;
            });
        const docs = mapToObject(pageStore);
        return { docs };
    }
}

function mapToObject<T>(map: Map<string, T>): { [key: string]: T } {
    const object: { [key: string]: T } = {};
    for (const [key, val] of map.entries()) {
        object[key] = val;
    }
    return object;
}
