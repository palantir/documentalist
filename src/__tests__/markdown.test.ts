/**
 * Copyright 2017-present Palantir Technologies, Inc. All rights reserved.
 * Licensed under the BSD-3 License as modified (the “License”); you may obtain
 * a copy of the license in the LICENSE and PATENTS files in the root of this
 * repository.
 */

import { Documentalist } from "../documentalist";
import { MarkdownPlugin } from "../plugins/markdown";

const TEST_MARKDOWN = `---
key: value
---

@# I'm special

## I'm regular

@othertag params
`;

const TEST_NAV = `@page test
`;

const TEST_FILES = [
    {
        path: "/whatever/other/test.md",
        read: () => TEST_MARKDOWN,
    },
    {
        path: "/who/cares/_nav.md",
        read: () => TEST_NAV,
    },
];

describe("MarkdownPlugin", () => {
    const dm = Documentalist.create().use(".md", new MarkdownPlugin());

    it("snapshot", async () => {
        const { pages } = await dm.documentFiles(TEST_FILES);
        expect(pages).toMatchSnapshot();
    });
});
