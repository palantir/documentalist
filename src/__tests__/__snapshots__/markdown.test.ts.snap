// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkdownPlugin snapshot 1`] = `
Object {
  "_nav": Object {
    "contents": Array [
      Object {
        "tag": "page",
        "value": "test",
      },
    ],
    "contentsRaw": "@page test",
    "metadata": Object {},
    "reference": "_nav",
    "route": "_nav",
    "title": "(untitled)",
  },
  "test": Object {
    "contents": Array [
      Object {
        "level": 1,
        "route": "test",
        "tag": "heading",
        "value": "I'm special",
      },
      "<h2 id=\\"i-m-regular\\">I&#39;m regular</h2>
",
      Object {
        "tag": "othertag",
        "value": "params",
      },
    ],
    "contentsRaw": "
@# I'm special

## I'm regular

@othertag params",
    "metadata": Object {
      "key": "value",
    },
    "reference": "test",
    "route": "test",
    "title": "I'm special",
  },
}
`;
